packages/server/lib/cloud/artifacts/upload_artifacts.ts

```diff
@@ -0,0 +1,205 @@
+    } catch (e) {
+      debug('Error creating video artifact: %O', e)
+    }
+  }
+
+  debug('screenshot metadata: %O', { screenshotUploadUrls, screenshots })
+  debug('found screenshot filenames: %o', screenshots)
+  if (screenshots?.length && screenshotUploadUrls?.length) {
+    const screenshotArtifacts = await createScreenshotArtifactBatch(screenshotUploadUrls, screenshots)
+
+    screenshotArtifacts.forEach((screenshot) => {
+      artifacts.push(screenshot)
+    })
+  }
+
+  try {
+    const protocolFilePath = protocolManager?.getArchivePath()
+
+    const protocolUploadUrl = captureUploadUrl || protocolCaptureMeta.url
+
+    debug('should add protocol artifact? %o, %o, %O', protocolFilePath, protocolUploadUrl, protocolManager)
+    if (protocolManager && protocolFilePath && protocolUploadUrl) {
+      artifacts.push(await createProtocolArtifact(protocolFilePath, protocolUploadUrl, protocolManager))
+    }
+  } catch (e) {
+    debug('Error creating protocol artifact: %O', e)
+  }
+
+  return artifacts
+}
+
+export const uploadArtifacts = async (options: UploadArtifactOptions) => {
+  const { protocolManager, protocolCaptureMeta, quiet, runId, instanceId, spec, platform, projectId } = options
+
+  const priority = {
+    'video': 0,
+    'screenshots': 1,
+    'protocol': 2,
+  }
+
+  const artifacts = (await extractArtifactsFromOptions(options)).sort((a, b) => {
+    return priority[a.reportKey] - priority[b.reportKey]
+  })
+
+  let uploadReport: UpdateInstanceArtifactsPayload
+
+  if (!quiet) {
+    logUploadManifest(artifacts, protocolCaptureMeta, protocolManager?.getFatalError())
+  }
+
+  debug('preparing to upload artifacts: %O', artifacts)
+
+  let stopUploadActivityOutput: () => void | undefined
+
+  if (!quiet && artifacts.length) {
+    stopUploadActivityOutput = beginUploadActivityOutput()
+  }
+
+  try {
+    const uploadResults = await Promise.all(artifacts.map((artifact) => artifact.upload())).finally(() => {
+      if (stopUploadActivityOutput) {
+        stopUploadActivityOutput()
+      }
+    })
+
+    /**
+     * upload report *stdout* is skipped for *some* fatal protocol runs, but we still
+     * want them reported to cloud.
+     */
+
+    if (!quiet && uploadResults.length) {
+      logUploadResults(uploadResults, protocolManager.getFatalError())
+    }
+
+    const protocolFatalError = protocolManager.getFatalError()
+
+    if (!uploadResults.find((result: ArtifactUploadResult) => {
+      return result.key === 'protocol'
+    }) && protocolFatalError) {
+      uploadResults.push(await composeProtocolErrorReportFromOptions(options))
+    }
+
+    uploadReport = uploadResults.reduce(toUploadReportPayload, { video: undefined, screenshots: [], protocol: undefined })
+  } catch (err) {
+    errors.warning('CLOUD_CANNOT_UPLOAD_ARTIFACTS', err)
+
+    return exception.create(err)
+  }
+
+  debug('checking for protocol errors', protocolManager?.hasErrors())
```

Comment:
```suggestion
  debug('checking for protocol errors', protocolManager.hasErrors())
```

Reply:
addressed in a6339bac2acb1bf0fa7fa4994aa80909b418e8ba